server/service/mdm.go

```diff
@@ -2618,130 +2619,143 @@ func (svc *Service) UploadMDMAppleVPPToken(ctx context.Context, token io.ReadSee
 		if errors.As(err, &vppErr) {
 			// Per https://developer.apple.com/documentation/devicemanagement/app_and_book_management/app_and_book_management_legacy/interpreting_error_codes
 			if vppErr.ErrorNumber == 9622 {
-				return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("token", "Invalid token. Please provide a valid content token from Apple Business Manager."))
+				return nil, ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("token", "Invalid token. Please provide a valid content token from Apple Business Manager."))
 			}
 		}
-		return ctxerr.Wrap(ctx, err, "validating VPP token with Apple")
+		return nil, ctxerr.Wrap(ctx, err, "validating VPP token with Apple")
 	}
 
 	data := fleet.VPPTokenData{
 		Token:    string(tokenBytes),
 		Location: locName,
 	}
 
-	dataBytes, err := json.Marshal(data)
+	tok, err := svc.ds.InsertVPPToken(ctx, &data, nil, fleet.NullTeamNoTeam)
 	if err != nil {
-		return ctxerr.Wrap(ctx, err, "creating VPP data object for storage")
-	}
-
-	err = svc.ds.ReplaceMDMConfigAssets(ctx, []fleet.MDMConfigAsset{
-		{Name: fleet.MDMAssetVPPTokenDeprecated, Value: dataBytes},
-	})
-	if err != nil {
-		return ctxerr.Wrap(ctx, err, "writing VPP token to db")
+		return nil, ctxerr.Wrap(ctx, err, "writing VPP token to db")
 	}
 
 	act := fleet.ActivityEnabledVPP{}
 	if err := svc.NewActivity(ctx, authz.UserFromContext(ctx), act); err != nil {
-		return ctxerr.Wrap(ctx, err, "create activity for upload VPP token")
+		return nil, ctxerr.Wrap(ctx, err, "create activity for upload VPP token")
 	}
 
-	return nil
+	return tok, nil
 }
 
-////////////////////////////////////////////////////////////////////////////////
-// GET /vpp
-////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////
+// PATCH /api/_version_/fleet/vpp_tokens/%d/teams //
+////////////////////////////////////////////////////
 
-type getMDMAppleVPPTokenRequest struct{}
+type patchVPPTokensTeamsRequest struct {
+	ID      uint    `url:"id"`
+	TeamIDs *[]uint `json:"team_id"`
```

Comment:
I don't think you need a pointer to the slice? The slice will be `nil` if the json is set to null: https://go.dev/play/p/E7mtuotBM7f.

Reply:
The reason I did this is that we need to differentiate between an empty slice and a `null` slice. I don't think you can detect the difference without a pointer? Aren't empty slices pointers to `nil`?

Reply:
No you can detect that, look at the playground link I posted - it unmarshals as a `nil` slice if the value is null or missing in the JSON, but it unmarshals to a non-nil empty slice if the value is an empty JSON array.

Reply:
Oh I didn't actually click the playground at first, cool and thanks!